pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DoubleGroupByResultHolder.java

```diff
@@ -102,12 +142,94 @@ public double getDoubleResult(int groupKey) {
    */
   @Override
   public void setValueForKey(int groupKey, double newValue) {
-    _resultArray[groupKey] = newValue;
+    if (_storageMode == StorageMode.ARRAY_STORAGE) {
+      _resultArray[groupKey] = newValue;
+    } else {
+      _resultMap.put(groupKey, newValue);
+      _priorityQueue.add(new IntDoublePair(groupKey, newValue));
+    }
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * @param groupKey Key for which to set the value
+   * @param newValue Value to set
+   */
   @Override
   public void setValueForKey(int groupKey, Object newValue) {
     throw new RuntimeException(
         "Unsupported method 'setValueForKey' (with Object param) for class " + getClass().getName());
   }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Keys with 'lowest' values (as per the sort order) are trimmed away to reduce
+   * the size to _maxCapacity.
+   *
+   * @return Array of keys that were trimmed out.
+   */
+  @Override
+  public int[] trimResults() {
+    if (_storageMode == StorageMode.ARRAY_STORAGE) {
+      return EMPTY_ARRAY; // Still in array mode, trimming has not kicked in yet.
+    }
+
+    int size = _resultMap.size();
+    if (size <= _maxCapacity) {
+      return EMPTY_ARRAY;
+    }
+
+    int numKeysToRemove = size - _maxCapacity;
```

Comment:
Yes, size of priority queue would be the total number of keys (renamed the variable name to reflect the same).